execution/exchange/duplicate_label.go

```diff
@@ -0,0 +1,125 @@
+
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/thanos-io/promql-engine/execution/model"
+	"github.com/thanos-io/promql-engine/extlabels"
+	"github.com/thanos-io/promql-engine/query"
+)
+
+type duplicateLabelCheckOperator struct {
+	model.OperatorTelemetry
+
+	once sync.Once
+	next model.VectorOperator
+
+	checkDuplicates func([]model.StepVector) error
+}
+
+func NewDuplicateLabelCheck(next model.VectorOperator, opts *query.Options) model.VectorOperator {
+	return &duplicateLabelCheckOperator{
+		OperatorTelemetry: model.NewTelemetry("[duplicateLabelCheck]", opts.EnableAnalysis),
+		next:              next,
+	}
+}
+
+func (d *duplicateLabelCheckOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	if err := d.initOnce(ctx); err != nil {
+		return nil, err
+	}
+	start := time.Now()
+	defer func() { d.AddExecutionTimeTaken(time.Since(start)) }()
+
+	in, err := d.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if in == nil {
+		return nil, nil
+	}
+
+	if err := d.checkDuplicates(in); err != nil {
+		return nil, err
+	}
+	return in, nil
+}
+
+func (d *duplicateLabelCheckOperator) Series(ctx context.Context) ([]labels.Labels, error) {
+	return d.next.Series(ctx)
+}
+
+func (d *duplicateLabelCheckOperator) GetPool() *model.VectorPool {
+	return d.next.GetPool()
+}
+
+func (d *duplicateLabelCheckOperator) Explain() (me string, next []model.VectorOperator) {
+	return d.next.Explain()
+}
+
+func (d *duplicateLabelCheckOperator) initOnce(ctx context.Context) error {
+	var err error
+	d.once.Do(func() {
+		series, seriesErr := d.next.Series(ctx)
+		if err != nil {
+			err = seriesErr
+			return
+		}
+		d.checkDuplicates = checkDuplicatesFn(series)
+	})
+	return err
+}
+
+// checkDuplicatesFn returns a function that checks a batch of step vectors for duplicate labels
+// that are populated at the same time. We do that by checking the Series response of the nested
+// operator for duplicates and note their seriesID. The seriesIDs that correspond to same labels
+// are then checked for being set at the same timestamp. We do that with a bitmask right now as
+// follows: if p and p' were set at batch index i we set the i-th bit in a bitmask in a helper array
+// for index p and p'. Later we OR them to see if they were set at the same time.
+// This method is NOT safe to be used with batching since it assumes that we have the whole time
+// slice at hand. It is safe for stepBatch sizes up to 64.
+func checkDuplicatesFn(series []labels.Labels) func([]model.StepVector) error {
+	type pair struct{ a, b int }
+
+	m := make(map[uint64]int, len(series))
+	p := make([]pair, 0)
+	c := make([]uint64, len(series))
+
+	for i := range series {
+		h := series[i].Hash()
+		if j, ok := m[h]; ok {
+			p = append(p, pair{a: i, b: j})
+		} else {
+			m[h] = i
+		}
+	}
+
+	// Shortcut: if we know there are no collision pairs we can just not check for duplicates
+	if len(p) == 0 {
+		return func(sv []model.StepVector) error { return nil }
```

Comment:
Let's make this its own function and give it a name. Besides making a new allocation each time, it's also hard to see how the no-op path is implemented.

Reply:
Should we just replace it by a Not nil check in Next? And only set the function if there are collision candidates?

Reply:
I am not a big fan of `nil` because you always have to worry about it.

Reply:
Ok

Reply:
Ok changed it up to work without closures!